internal/rbd/qos.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"strconv"
+
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+)
+
+const (
+	// Qos parameters name of StorageClass.
+	baseReadIops            = "BaseReadIops"
+	baseWriteIops           = "BaseWriteIops"
+	baseReadBytesPerSecond  = "BaseReadBytesPerSecond"
+	baseWriteBytesPerSecond = "BaseWriteBytesPerSecond"
+	readIopsPerGB           = "ReadIopsPerGB"
+	writeIopsPerGB          = "WriteIopsPerGB"
+	readBpsPerGB            = "ReadBpsPerGB"
+	writeBpsPerGB           = "WriteBpsPerGB"
+	baseVolSizeBytes        = "BaseVolSizeBytes"
+
+	// Qos type name of rbd image.
+	readIopsLimit      = "rbd_qos_read_iops_limit"
+	writeIopsLimit     = "rbd_qos_write_iops_limit"
+	readBpsLimit       = "rbd_qos_read_bps_limit"
+	writeBpsLimit      = "rbd_qos_write_bps_limit"
+	metadataConfPrefix = "conf_"
+
+	// The params use to calc qos based on capacity.
+	baseQosReadIopsLimit  = "rbd_base_qos_read_iops_limit"
+	baseQosWriteIopsLimit = "rbd_base_qos_write_iops_limit"
+	baseQosReadBpsLimit   = "rbd_base_qos_read_bps_limit"
+	baseQosWriteBpsLimit  = "rbd_base_qos_write_bps_limit"
+	readIopsPerGBLimit    = "rbd_read_iops_per_gb_limit"
+	writeIopsPerGBLimit   = "rbd_write_iops_per_gb_limit"
+	readBpsPerGBLimit     = "rbd_read_bps_per_gb_limit"
+	writeBpsPerGBLimit    = "rbd_write_bps_per_gb_limit"
+	baseQosVolSize        = "rbd_base_qos_vol_size"
+)
+
+type qosSpec struct {
+	baseLimitType  string
+	baseLimit      string
+	perGBLimitType string
+	perGBLimit     string
+	provide        bool
+}
+
+func parseQosParams(
+	scParams map[string]string,
+) map[string]*qosSpec {
+	rbdQosParameters := map[string]*qosSpec{
+		baseReadIops:            {readIopsLimit, "", readIopsPerGB, "", false},
+		baseWriteIops:           {writeIopsLimit, "", writeIopsPerGB, "", false},
+		baseReadBytesPerSecond:  {readBpsLimit, "", readBpsPerGB, "", false},
+		baseWriteBytesPerSecond: {writeBpsLimit, "", writeBpsPerGB, "", false},
+	}
+	for k, v := range scParams {
+		if qos, ok := rbdQosParameters[k]; ok && v != "" {
+			qos.baseLimit = v
+			qos.provide = true
```

Comment:
can we rename `provide` to something more accurate? like present etc?

Reply:
updated